### PR TITLE
txn support IS_EXIST and NOT_EXIST compare

### PR DESCRIPTION
**Pull request by yylt**
txn  support IS_EXIST and NOT_EXIST compare 

Fix https://github.com/etcd-io/etcd/issues/10566

cc @purpleidea

### CODE REVIEW

**Comment by purpleidea** [comment]
wow, awesome! I'd recommend we add a test with this patch to keep it stable! Also, do we want to guarantee the behaviour of checking if a value when that value doesn't exist? Should it be an error instead?

**Comment by jingyih** [comment]
Could you provide an example usage? Is this trying to decide whether a key exist? I am trying to understand the difference between this and something like: `clientv3.Compare(clientv3.ModRevision("key"), ">", 0)`

**Comment by yylt** [comment]
@jingyih yes, this trying to decide whether a key exist or not exist. I think `ModRevision` may be used as Optimistic lock mostly .

File: clientv3/example_kv_test.go [hunk 1]
```diff
@@ -259,6 +259,43 @@ func ExampleKV_txn() {
...
+		log.Fatal(err)
+	}
+	defer cli.Close()
+
+	kvc := clientv3.NewKV(cli)
+
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	_, err = kvc.Txn(ctx).
+	// txn value comparisons are lexical
+		If(clientv3.CompareExist(clientv3.Value("key"), "ne")).
```
Comment (jingyih) [inline clientv3/example_kv_test.go]: Sorry I am still confused. If we replace this line with the following line, isn't it the same?
`If(clientv3.Compare(clientv3.ModRevision("key"), "=", 0))`

Comment (jingyih) [reply clientv3/example_kv_test.go]: BTW, I don't think this PR solves the original issue #10566. The original issue is that when we use clientv3.Compare(clientv3.Value(key), "=", value), the compare always return `false` if key does not exist, regardless of "=" or "!=".

Comment (yylt) [reply clientv3/example_kv_test.go]: Thank you for your suggestions.

1: There are the same, i just think `ModRevision` make user realize this API should be use as Optimistic lock in mvcc storage


2: `clientv3.Compare(clientv3.Value(key), "=", value)` always return false, the reason just is 

https://github.com/etcd-io/etcd/blob/dc6885d73f71d4f7e77007fcc794f36e09fa5ecb/etcdserver/apply.go#L449

This PR had delete these codes.
I am also confused why add these codes, and think value compare result should be decided by bytes.Compare

**Comment by purpleidea** [comment]
So there's this: https://godoc.org/github.com/coreos/etcd/clientv3/clientv3util

Secondly, I think this relates to the whole design of what should happen if you try to read the value of a key that doesn't exist. I think this should propagate upwards as an error. This is the equivalent of a nil in golang. I can explain this better if it would help.

**Comment by yylt** [comment]
Wow, I will close this PR
Thank your explanation @jingyih @purpleidea

File: clientv3/example_kv_test.go (hunk 1 above)
Comment (jingyih) [reply clientv3/example_kv_test.go]: > Thank you for your suggestions.
> 
> 1: There are the same, i just think `ModRevision` make user realize this API should be use as Optimistic lock in mvcc storage

Thanks for confirming. I think we should not break API to implement a feature that already exists.

> 2: `clientv3.Compare(clientv3.Value(key), "=", value)` always return false, the reason just is
> 
> https://github.com/etcd-io/etcd/blob/dc6885d73f71d4f7e77007fcc794f36e09fa5ecb/etcdserver/apply.go#L449
> 
> This PR had delete these codes.
> I am also confused why add these codes, and think value compare result should be decided by bytes.Compare

Sorry I missed the deleting part. Removing the code block means treating non-exist-key's value as an empty string (or its byte presentation). This is probably fine, but we should more carefully think through this.

**Comment by jingyih** [comment]
@purpleidea Thanks for the link. I agree the behavior should be carefully designed. The discussion can be continued in the original issue.

**Comment by jingyih** [comment]
@yylt Thanks for trying to fix this!

**Comment by purpleidea** [comment]
@yylt I think there was some promise in the PR, in particular when I previously read through the code, I noticed the: https://github.com/etcd-io/etcd/pull/10769/files#diff-f141df18bc5c4d2e821b7d9dfb484f95L449 which you removed was probably a bug which was masking weird key not exists handling as discussed in https://github.com/etcd-io/etcd/issues/10566
So I think you're on the right track.

Perhaps you'd like to take what you learned digging through this part of the code and look at https://github.com/etcd-io/etcd/issues/10571 ?

**Comment by yylt** [comment]
@purpleidea  yes, I can try the feature https://github.com/etcd-io/etcd/issues/10571

**Pull request closed without merging**